packages/connect/src/core/onCallFirmwareUpdate.ts

```diff
@@ -0,0 +1,508 @@
+            },
+            context: { deviceList, device, log, postMessage },
+        });
+    }
+
+    if (!deviceInitiallyConnectedInBootloader && manual) {
+        await waitForDisconnectedDevice({
+            params: { manual },
+            context: { deviceList, device, postMessage, log },
+        });
+    }
+
+    if (!params.binary && device.firmwareRelease.intermediaryVersion) {
+        log.debug('onCallFirmwareUpdate', 'installing intermediary');
+        reconnectedDevice = await waitForReconnectedDevice({
+            params: {
+                bootloader: true,
+                manual,
+                // todo: check if correct?
+                confirmOnDevice: true,
+            },
+            context: { deviceList, device, log, postMessage },
+        });
+
+        // note: fw major_version 1 requires calling initialize before calling FirmwareErase. Without it device would not respond
+        await reconnectedDevice.initialize(false, false);
+
+        await uploadFirmware(
+            reconnectedDevice.getCommands().typedCall.bind(reconnectedDevice.getCommands()),
+            postMessage,
+            reconnectedDevice,
+            { payload: binary },
+        );
+
+        await reconnectedDevice.release();
+
+        await waitForDisconnectedDevice({
+            params: { manual },
+            context: { deviceList, log, device: reconnectedDevice, postMessage },
+        });
+        reconnectedDevice = await waitForReconnectedDevice({
+            params: { bootloader: true, manual, confirmOnDevice: false },
+            context: { deviceList, device: reconnectedDevice, log, postMessage },
+        });
+
+        binary = await getBinaryHelper(reconnectedDevice, params, log, postMessage);
+
+        // note: fw major_version 1 requires calling initialize before calling FirmwareErase. Without it device would not respond
+        await reconnectedDevice.initialize(false, false);
+
+        await uploadFirmware(
+            reconnectedDevice.getCommands().typedCall.bind(reconnectedDevice.getCommands()),
+            postMessage,
+            reconnectedDevice,
+            { payload: stripped },
+        );
+
+        await reconnectedDevice.release();
+    } else {
+        reconnectedDevice = await waitForReconnectedDevice({
+            params: { bootloader: true, manual, confirmOnDevice: true },
+            context: { deviceList, device, log, postMessage },
+        });
+
+        // note: fw major_version 1 requires calling initialize before calling FirmwareErase. Without it device would not respond
+        await reconnectedDevice.initialize(false, false);
+
+        await uploadFirmware(
+            reconnectedDevice.getCommands().typedCall.bind(reconnectedDevice.getCommands()),
+            postMessage,
+            reconnectedDevice,
+            { payload: shouldStripFwHeaders(device.features) ? stripped : binary },
+        );
+
+        await reconnectedDevice.release();
+    }
```

Comment:
This `else` branch is suspiciously similar (apart from what's sent as a fw binary) to the beginning of its `if` branch. Could it be unified like this? `if (cond) { A; B; } else { A; }` -> `A; if (cond) { B; }`